### PR TITLE
refactor: analysis views

### PR DESCRIPTION
**Pull request by derrabauke**
## Contents
* d77f2c2 fix(filter-sidebar): use correct QPs for filter sidebar
* b124f57 test(analysis-edit): can not click disabled button via click
* d87cdcf refactor(filter-sidebar): repair and octanify
* 9c855b4 test(analysis): correct selectors and fix setup
* ba31501 test(tracking): setup mock service
* 4646420 fix(analysis): unintended queryParams reset when editing reports
* 14bd695 refactor(analysis): query parameter setup
* 4be6bdf refactor(analysis): octanify analysis routes
* bcf6e18 chore: ignore window undef error
* 304f730 refactor(tracking): prevent use of observers
* 3514072 fix(tracking): show correct task duration sum

---

### CODE REVIEW

File: app/analysis/index/controller.js [hunk 1]
```diff
@@ -84,27 +84,27 @@ export const AnalysisQueryParams = new QueryParams({
     deserialize: deserializeMoment,
   },
   review: {
-    defaultValue: "",
+    defaultValue: null,
```
Comment (derrabauke) [inline app/analysis/index/controller.js:87]: Without *nulling* those QPs would also appear in the URL. Which is not intended. :(

File: app/analysis/edit/controller.js [hunk 2]
```diff
@@ -87,58 +101,46 @@ export default Controller.extend(AnalysisEditQueryParams.Mixin, {
...
+  get _customer() {
+    const id = get(this, "intersectionModel.customer.id");
     return id && this.store.peekRecord("customer", id);
-  }),
+  }
 
-  _project: computed("intersectionModel.project.id", "store", function () {
-    const id = this.get("intersectionModel.project.id");
+  get _project() {
+    const id = get(this, "intersectionModel.project.id");
```
Comment (velrest) [inline app/analysis/edit/controller.js:112]: `get` necessary or just easier refactoring?

File: app/analysis/edit/template.hbs [hunk 3]
```diff
@@ -95,7 +90,7 @@
                     </div>
                   </f.input>
 
-                  <f.input data-test-comment @name="comment" as |fi|>
```
Comment (velrest) [inline app/analysis/edit/template.hbs:98]: is the selector just not needed? Otherwise failing tests incoming

File: app/analysis/edit/controller.js (hunk 2 above)
Comment (derrabauke) [reply app/analysis/edit/controller.js:112]: Necessary, since we are accessing a proxy.

File: app/analysis/index/controller.js [hunk 4]
```diff
@@ -35,228 +35,202 @@ const deserializeMoment = (momentString) =>
...
-  _scrollOffset: 0,
+  exportLimit = config.APP.EXPORT_LIMIT;
 
-  init(...args) {
-    this._super(...args);
+  @service session;
+  @service store;
+  @service router;
+  @service notify;
+  @service can;
```
Comment (velrest) [inline app/analysis/index/controller.js:184]: Nitpick, would probably move services to start of class definition since this is kind of the unspoken norm i think.

File: app/analysis/index/controller.js [hunk 5]
```diff
@@ -274,64 +248,67 @@ const AnalysisController = Controller.extend(AnalysisQueryParams.Mixin, {
...
-        number: this._lastPage + 1,
-        size: 20,
-      },
-      ...params,
-      include: "task,task.project,task.project.customer,user",
-    });
+    if (this._canLoadMore) {
+      const data = yield this.store.query("report", {
+        page: {
+          number: this._canLoadMore ? this._lastPage + 1 : this._lastPage,
```
Comment (velrest) [inline app/analysis/index/controller.js:262]: Ternary check here is unnecessary since your already in an `if(this.canLoadMore)`

File: app/analysis/index/controller.js [hunk 6]
```diff
@@ -368,117 +345,113 @@ const AnalysisController = Controller.extend(AnalysisQueryParams.Mixin, {
...
-
-        // ignore since we can't really test this..
-        /* istanbul ignore next */
-        if (!Ember.testing) {
-          download(file, filename, file.type);
-        }
-
-        notify.success("File was downloaded");
-      } catch (e) {
-        /* istanbul ignore next */
```
Comment (velrest) [inline app/analysis/index/controller.js:439]: Are we actually not using codecoverage anymore in timed?

File: app/application/template.hbs [hunk 7]
```diff
@@ -1,6 +1,6 @@
-{{outlet}}
-
 <EmberNotify @messageStyle="bootstrap" />
 <SyModalTarget />
 
 <div id="filter-sidebar-target"></div>
+
+{{outlet}}
```
Comment (velrest) [inline app/application/template.hbs:6]: Just for the future to keep in mind if we have funky rendering issues, (same as in wigl) this might be the cause

File: app/components/filter-sidebar/filter/template.hbs [hunk 8]
```diff
@@ -1,46 +1,61 @@
...
+          {{get opt @labelPath}}
+        </button>
+        {{yield}}
+      {{/each}}
+    </div>
+  {{else if (eq @type "date")}}
+    <SyDatepicker
+      @value={{@selected}}
+      @onChange={{optional @onChange}}
+      @placeholder={{(or @placeholder)}}
```
Comment (velrest) [inline app/components/filter-sidebar/filter/template.hbs:19]: Does `@placeholder` behave weirdly when a falsy value is passed or why not just `{{@placeholder}}`?

File: app/components/filter-sidebar/template.hbs [hunk 9]
```diff
@@ -1,39 +1,51 @@
...
+      class="filter-sidebar-toggle"
+      type="button"
+      {{on "click" (toggle "visible" this)}}
     >
       <span class="filter-sidebar-toggle-background"></span>
       <span class="filter-sidebar-toggle-icon">
-        {{fa-icon 'sliders'}}
-        {{#if this.appliedCount}}
-          <span class="filter-sidebar-toggle-icon-badge">{{this.appliedCount}}</span>
+        {{fa-icon "sliders"}}
```
Comment (velrest) [inline app/components/filter-sidebar/template.hbs:36]: Modifier or Component?

File: app/components/sy-modal/template.hbs [hunk 10]
```diff
@@ -1,12 +1,16 @@
...
-      {{yield (hash
-        header = (component 'sy-modal/header' close=this.close)
-        body   = (component 'sy-modal/body'   close=this.close)
-        footer = (component 'sy-modal/footer' close=this.close)
-        close  = (this.close)
-      )}}
-    </div>
-  </SyModal::Overlay>
-</EmberWormhole>
+{{#if @visible}}
```
Comment (velrest) [inline app/components/sy-modal/template.hbs:1]: Why is this needed? SyOverlay already seems to have an @visible

File: app/index/controller.js [hunk 11]
```diff
@@ -86,16 +100,33 @@ export default class IndexController extends Controller {
...
+    // See activitySum getter.
+    scheduleOnce("afterRender", this, "_storeLastActivitySum");
+
     return duration;
   }
 
+  _storeLastActivitySum() {
+    this._lastActivitySum = this.activitySum;
+  }
+
```
Comment (velrest) [inline app/index/controller.js:129]: My brain tells me there would probably be a better way to solve this but sadly lacks the power to tell me the solution.

File: app/utils/url.js [hunk 12]
```diff
@@ -7,5 +7,6 @@ export const cleanParams = (params) =>
 
 export const toQueryString = (params) =>
   Object.keys(params)
+    .filter((key) => !!params[key])
```
Comment (velrest) [inline app/utils/url.js:10]: IMO just wrapping it in `Boolean` is easier to read.

File: app/analysis/index/controller.js (hunk 6 above)
Comment (derrabauke) [reply app/analysis/index/controller.js:439]: It's currently deactivated. Since when: don't know. But we should activate it again soon. 

I deleted already some of those comments, since they annoyed me, but this was stupid I guess, since they are markers for codecov...

File: app/components/filter-sidebar/filter/template.hbs (hunk 8 above)
Comment (derrabauke) [reply app/components/filter-sidebar/filter/template.hbs:19]: Shit, that's a debugging artifact.

File: app/components/filter-sidebar/template.hbs (hunk 9 above)
Comment (derrabauke) [reply app/components/filter-sidebar/template.hbs:36]: `component` --> `<FaIcon @icon="sliders" />

File: app/index/controller.js (hunk 11 above)
Comment (derrabauke) [reply app/index/controller.js:129]: Mine got exhausted by this code as well... this whole mechanism is weird. I just tried to get it running again.

File: app/analysis/index/controller.js [hunk 13]
```diff
@@ -35,228 +35,202 @@ const deserializeMoment = (momentString) =>
...
-    defaultValue: null,
+    defaultValue: undefined,
     replace: true,
     refresh: true,
     serialize: serializeMoment,
     deserialize: deserializeMoment,
   },
   toDate: {
-    defaultValue: null,
+    defaultValue: undefined,
```
[the remark this answers is not in the document]
Comment (derrabauke) [reply app/analysis/index/controller.js:80]: `null` as default value leads to a `change` in the this.queryParamsState Object. And the change indicated is `null` -> `undefined`.  So I assume the defaultValue for query params if not defined should be `undefined` instead of `null`.

File: app/components/sy-modal/template.hbs (hunk 10 above)
Comment (derrabauke) [reply app/components/sy-modal/template.hbs:1]: The modal was still hanging in the DOM even though it was invisible. This way it will get cleaned properly. The component will get refactored for ember v4 anyways. :shrug:

**Review by velrest — APPROVED** [review]
